### PR TITLE
Use Bolt transport connections for execution

### PR DESCRIPTION
**Pull request by MikaelSmith**
Implement support for SSH and WinRM using Bolt's transport connections. This gives us consistent behavior with Bolt for those transports.

Only implements SSH and WinRM because Docker can use the Wash Docker plugin, local is trivial, and remote is not really usable. PCP I hope to implement later, but the lack of streaming responses will require more work.

### CODE REVIEW

File: bolt.rb [hunk 1]
```diff
@@ -117,36 +117,55 @@ class Target < Wash::Entry
...
+      raise "#{transport} unsupported"
+    end
+
+    begin
+      connection.connect
+      # Returns exit code
+      connection.execute(command, stdin: opts[:stdin] ? $stdin : nil)
+    ensure
+      begin
+        connection&.disconnect
```
Comment (ekinanp) [inline bolt.rb:167]: Should this be in a SIGTERM handler as well?

File: transport_ssh.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
+require 'bolt/transport/base'
+require 'bolt/transport/sudoable'
+require 'bolt/transport/ssh/connection'
+require 'concurrent/atomic/atomic_reference'
+require 'concurrent/promises'
+
+class BoltSSH < Bolt::Transport::SSH::Connection
+  # override execute so it streams output and returns the exit code
+  def execute(command, stdin: nil)
```
Comment (ekinanp) [inline transport_ssh.rb:12]: Some of this code seems copy-pasted from https://github.com/puppetlabs/bolt/blob/master/lib/bolt/transport/ssh/connection.rb#L224 ? If so, might be helpful to link that as a reference.

File: bolt.rb [hunk 3]
```diff
@@ -117,36 +117,55 @@ class Target < Wash::Entry
...
+      require_relative 'transport_winrm.rb'
+      connection = BoltWinRM.new(target, logger)
+    else
+      raise "#{transport} unsupported"
+    end
+
+    begin
+      connection.connect
+      # Returns exit code
+      connection.execute(command, stdin: opts[:stdin] ? $stdin : nil)
```
Comment (ekinanp) [inline bolt.rb]: Nit: Why not `stdin: opts[:stdin] || nil` ?`opts[:stdin]` should contain `$stdin`.

File: transport_ssh.rb (hunk 2 above)
Comment (ekinanp) [reply transport_ssh.rb:12]: You know, given all the code here, maybe breaking consistency isn't so bad. If I was reading this code, I'd be happy seeing something like
```
if supports_ssh
  // Implementing exec for SSH sucks, so we're just going to use Wash's
  // transport package to do it for us. Unfortunately no support exists for
  // WinRM, so we'll have to implement our own version in exec.
  transport :ssh, ...
end
```

**Review by ekinanp — COMMENTED** [review]
Looks good in general, just some small comments re: code.

Have a few questions though:

* How hard would it be to add a hacky `exec` for the Docker transport? Initial idea is an `ExecDocker` method in the `transport` package and then using the `transport` feature. I get that breaks consistency, but I think expecting people to use the Docker plugin is the type of move that could introduce "pain for no good reason", unless doing exec for Docker is really hard and something we don't have time for given other priorities. By pain, I mean that it not only breaks the Boltwash interactive experience, but it may also require some extra, frustrating effort for a user. For example, it could lead to something like

> What was this target's raw container ID? How did I figure that out? Oh maybe `meta` has that. Ok there it is. Now to the Docker plugin now.. Navigating through the "fs".. Wait maybe my issue's in another target. Ok back to Boltwash, where was I again? Ohh no, I guess it's not in this target. Maybe it's in this other container's logs... Hmm.. Let me cd into it.. Ugh, right. I need to go to the Docker plugin to browse this thing after I use `meta` to figure out its raw container ID, wow so much for the "we reuse your hierarchy" crap in the plugin's README

All it takes is one bad experience to piss someone off, but I could also be paranoid about this.

* You still planning on adding WinRM support to the `transport` package tomorrow/next week, or leaving it alone for now? Asking out of curiosity because I'm fine with what you've got here. Surprised at how much simpler it is than SSH though.

File: bolt.rb (hunk 1 above)
Comment (MikaelSmith) [reply bolt.rb:167]: Ensure handles that.

File: transport_ssh.rb (hunk 2 above)
Comment (MikaelSmith) [reply transport_ssh.rb:12]: It's not that much code, and it is pretty much copied from Bolt with minor changes.

File: bolt.rb (hunk 3 above)
Comment (MikaelSmith) [reply bolt.rb]: Oh, forgot it gets switched from a boolean to $stdin. It's not well documented in wash-ruby.

**Comment by MikaelSmith** [comment]
It's probably not too bad to support Docker. I'll add it.

**Comment by MikaelSmith** [comment]
Added a rough plan for Windows target support to https://github.com/puppetlabs/wash/issues/699.

File: transport_ssh.rb (hunk 2 above)
Comment (ekinanp) [reply transport_ssh.rb:12]: Ok, fair enough.

**Comment by MikaelSmith** [comment]
Ok, this is ready again.

File: transport_docker.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    command_options << '--tty' if target.options['tty']
+    command_options << container_id
+    command_options.concat(command)
+
+    env_hash = {}
+    # Set the DOCKER_HOST if we are using a non-default service-url
+    env_hash['DOCKER_HOST'] = @docker_host unless @docker_host.nil?
+
+    in_r, in_w = IO.pipe
+    in_w.sync = true
```
Comment (ekinanp) [inline transport_docker.rb:29]: What part of this streams to stdout/stderr?

**Review by ekinanp — APPROVED** [review]
Looks good, just had one question.

File: transport_docker.rb (hunk 4 above)
Comment (MikaelSmith) [reply transport_docker.rb:29]: Spawn's default behavior is to stream to inherit file descriptors (stdin/stdout/stderr) from the parent.

Comment (ekinanp) [reply transport_docker.rb:29]: Ah ok. Thanks

**Pull request merged**